chia/consensus/block_body_validation.py

```diff
@@ -353,35 +356,53 @@ async def validate_block_body(
             )
             removal_coin_records[new_unspent.name] = new_unspent
         else:
-            unspent = await coin_store.get_coin_record(rem)
-            if unspent is not None and unspent.confirmed_block_index <= fork_h:
-                # Spending something in the current chain, confirmed before fork
-                # (We ignore all coins confirmed after fork)
-                if unspent.spent == 1 and unspent.spent_block_index <= fork_h:
-                    # Check for coins spent in an ancestor block
-                    return Err.DOUBLE_SPEND, None
-                removal_coin_records[unspent.name] = unspent
-            else:
-                # This coin is not in the current heaviest chain, so it must be in the fork
-                if rem not in additions_since_fork:
-                    # Check for spending a coin that does not exist in this fork
-                    log.error(f"Err.UNKNOWN_UNSPENT: COIN ID: {rem} NPC RESULT: {npc_result}")
-                    return Err.UNKNOWN_UNSPENT, None
-                new_coin, confirmed_height, confirmed_timestamp = additions_since_fork[rem]
-                new_coin_record: CoinRecord = CoinRecord(
-                    new_coin,
-                    confirmed_height,
-                    uint32(0),
-                    False,
-                    confirmed_timestamp,
-                )
-                removal_coin_records[new_coin_record.name] = new_coin_record
-
             # This check applies to both coins created before fork (pulled from coin_store),
             # and coins created after fork (additions_since_fork)
             if rem in removals_since_fork:
                 # This coin was spent in the fork
                 return Err.DOUBLE_SPEND_IN_FORK, None
+            removals_from_db.append(rem)
+
+    unspent_records = await coin_store.get_coin_records(removals_from_db)
+
+    # some coin spends we need to ensure exist in the fork branch. Both coins we
+    # can't find in the DB, but also coins that were spent after the fork point
+    look_in_fork: List[bytes32] = []
+    for unspent in unspent_records:
+        if unspent.confirmed_block_index <= fork_h:
+            # Spending something in the current chain, confirmed before fork
+            # (We ignore all coins confirmed after fork)
+            if unspent.spent == 1 and unspent.spent_block_index <= fork_h:
+                # Check for coins spent in an ancestor block
+                return Err.DOUBLE_SPEND, None
+            removal_coin_records[unspent.name] = unspent
+        else:
+            look_in_fork.append(unspent.name)
+
+    if len(unspent_records) != len(removals_from_db):
+        # some coins could not be found in the DB. We need to find out which
+        # ones and look for them in additions_since_fork
+        found = set([u.name for u in unspent_records])
+        for rem in removals_from_db:
```

Comment:
do we have test coverage of all these cases? I'm scared one of these cases might be missing in tests. IIRC i tested all of the cases before, but we might want to check coverage here just in case

Reply:
all the cases in my patch are exercised by the tests. We can't necessarily conclude that all cases that we need to handle are covered by the tests though.

I count these different cases:

1. We spend a coin that was confirmed before or at the fork height (this is the normal, non-reorg case too)
2. We spend an ephemeral coin
3. We attempt to spend a coin that was spent in the fork we're considering (i.e. after the fork height)
4. We attempt to spend a coin that was spent before the fork height (i.e. a normal double spend, not reorg case)
5. We spend a coin that was confirmed in the fork we're considering (i.e. after the fork height. It won't be in the coin_store)
6. We attempt to spend a coin that doesn't exist in the main chain nor in the fork (this is `UNKNOWN_UNSPENT`)

case (2) is very well covered by the ephemeral coin test cases I added a while back.

We have these tests:
* [test_conditions](https://github.com/Chia-Network/chia-blockchain/blob/main/tests/blockchain/test_blockchain.py#L1758) (covers 1, 2)
* [double_spend_in_reorg](https://github.com/Chia-Network/chia-blockchain/blob/main/tests/blockchain/test_blockchain.py#L2653) (covers 1, 2, 3, 6)
* [minting_coin](https://github.com/Chia-Network/chia-blockchain/blob/main/tests/blockchain/test_blockchain.py#L2745) (covers 1)
* [reorg_flip_flip](https://github.com/Chia-Network/chia-blockchain/blob/main/tests/blockchain/test_blockchain.py#L3251) (covers 5)
* [chain_failed_rollback](https://github.com/Chia-Network/chia-blockchain/blob/main/tests/blockchain/test_blockchain.py#L3193) (covers 1)
* [reorg_stale_fork_height](https://github.com/Chia-Network/chia-blockchain/blob/main/tests/blockchain/test_blockchain.py#L3193) (covers 1, 5)
* [reorg_new_ref](https://github.com/Chia-Network/chia-blockchain/blob/main/tests/blockchain/test_blockchain.py#L3066) (covers 1)
* [get_header_blocks_in_range_tx_filter](https://github.com/Chia-Network/chia-blockchain/blob/main/tests/blockchain/test_blockchain.py#L3019) (covers 1)
* [reorg_transaction](https://github.com/Chia-Network/chia-blockchain/blob/main/tests/blockchain/test_blockchain.py#L2974) (covers 1)
* [double_spent_in_coin_store](https://github.com/Chia-Network/chia-blockchain/blob/main/tests/blockchain/test_blockchain.py#L2619) (covers 1, 4)